<!doctype html>
<html class="no-js" lang="en" data-content_root="../../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" />

    <!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>suvtk.features - suvtk 0.1</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">suvtk 0.1</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo only-light" src="../../_static/logo_light.svg" alt="Light Logo"/>
    <img class="sidebar-logo only-dark" src="../../_static/logo_dark.svg" alt="Dark Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">suvtk</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Commands</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../help/taxonomy.html">taxonomy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../help/features.html">features</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../help/virus-info.html">virus-info</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../help/co-occurrence.html">co-occurrence</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../help/gbk2tbl.html">gbk2tbl</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../help/comments.html">comments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../help/table2asn.html">table2asn</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Python API</span></p>
<ul>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../source/modules.html">suvtk</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of suvtk</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../source/suvtk.html">suvtk package</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <h1>Source code for suvtk.features</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">features.py</span>
<span class="sd">===========</span>

<span class="sd">This script processes input sequences to predict open reading frames (ORFs),</span>
<span class="sd">aligns the predicted protein sequences against a database, and generates feature</span>
<span class="sd">tables for submission to GenBank.</span>

<span class="sd">Functions</span>
<span class="sd">---------</span>
<span class="sd">validate_translation_table(ctx, param, value)</span>
<span class="sd">    Validate the given translation table.</span>

<span class="sd">calculate_coding_capacity(genes, seq_length)</span>
<span class="sd">    Calculate the total coding capacity for a list of genes.</span>

<span class="sd">find_orientation(genes)</span>
<span class="sd">    Determine the orientation of genes based on strand information.</span>

<span class="sd">predict_orfs(orf_finder, seq)</span>
<span class="sd">    Predict ORFs, compute coding capacity, and determine orientation.</span>

<span class="sd">features(fasta_file, output_path, database, transl_table, coding_complete, taxonomy, separate_files, threads)</span>
<span class="sd">    Main command to create feature tables for sequences.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># TODO: add count for sequences without ORF prediction</span>
<span class="c1"># TODO: mmseqs log to file for clarity</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">shutil</span>

<span class="kn">import</span> <span class="nn">Bio.SeqIO</span>
<span class="kn">import</span> <span class="nn">click</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pyrodigal_gv</span>
<span class="kn">import</span> <span class="nn">taxopy</span>
<span class="kn">from</span> <span class="nn">Bio.SeqIO</span> <span class="kn">import</span> <span class="n">write</span>

<span class="kn">from</span> <span class="nn">suvtk</span> <span class="kn">import</span> <span class="n">utils</span>

<span class="c1"># Define the valid genetic codes based on NCBI</span>
<span class="n">VALID_GENETIC_CODES</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span> <span class="o">|</span> <span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">17</span><span class="p">))</span> <span class="o">|</span> <span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">21</span><span class="p">,</span> <span class="mi">32</span><span class="p">))</span>


<div class="viewcode-block" id="validate_translation_table">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.validate_translation_table">[docs]</a>
<span class="k">def</span> <span class="nf">validate_translation_table</span><span class="p">(</span><span class="n">ctx</span><span class="p">,</span> <span class="n">param</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Validate that the given translation table is one of the valid genetic codes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ctx : click.Context</span>
<span class="sd">        The Click context object. Unused.</span>
<span class="sd">    param : click.Parameter</span>
<span class="sd">        The parameter object. Unused.</span>
<span class="sd">    value : int</span>
<span class="sd">        The given translation table.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        The given translation table if it is valid.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    click.BadParameter</span>
<span class="sd">        If the given translation table is not valid.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">value</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">VALID_GENETIC_CODES</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">click</span><span class="o">.</span><span class="n">BadParameter</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Invalid translation table. Must be one of </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="n">VALID_GENETIC_CODES</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">value</span></div>



<div class="viewcode-block" id="calculate_coding_capacity">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.calculate_coding_capacity">[docs]</a>
<span class="k">def</span> <span class="nf">calculate_coding_capacity</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">seq_length</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the total coding capacity for a list of genes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genes : list</span>
<span class="sd">        A list of gene objects.</span>
<span class="sd">    seq_length : int</span>
<span class="sd">        The length of the sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        The total coding capacity.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">((</span><span class="n">gene</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">gene</span><span class="o">.</span><span class="n">begin</span><span class="p">)</span> <span class="o">/</span> <span class="n">seq_length</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">)</span></div>



<div class="viewcode-block" id="find_orientation">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.find_orientation">[docs]</a>
<span class="k">def</span> <span class="nf">find_orientation</span><span class="p">(</span><span class="n">genes</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the sum of the strand orientations for a list of genes.</span>
<span class="sd">    If the sum is zero, return the orientation of the largest gene.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genes : list</span>
<span class="sd">        A list of gene objects, each having &#39;strand&#39;, &#39;begin&#39;, and &#39;end&#39; attributes.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        The sum of strand orientations across all genes, or the orientation of the largest gene if the sum is zero.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">orientation_sum</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">strand</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">orientation_sum</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">genes</span><span class="p">:</span>
        <span class="c1"># Find the largest gene based on absolute length (|end - begin|)</span>
        <span class="n">largest_gene</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">abs</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">gene</span><span class="o">.</span><span class="n">begin</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">largest_gene</span><span class="o">.</span><span class="n">strand</span>

    <span class="k">return</span> <span class="n">orientation_sum</span></div>



<div class="viewcode-block" id="extract_gene_results">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.extract_gene_results">[docs]</a>
<span class="k">def</span> <span class="nf">extract_gene_results</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">record_id</span><span class="p">,</span> <span class="n">seq_length</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract gene prediction results for a sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genes : list</span>
<span class="sd">        A list of gene objects.</span>
<span class="sd">    record_id : str</span>
<span class="sd">        The ID of the sequence record.</span>
<span class="sd">    seq_length : int</span>
<span class="sd">        The length of the sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list</span>
<span class="sd">        A list of gene prediction results.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="n">record_id</span><span class="p">,</span>
            <span class="n">seq_length</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">record_id</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">begin</span><span class="p">,</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">start_node</span><span class="o">.</span><span class="n">type</span><span class="p">,</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">partial_begin</span><span class="p">,</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">partial_end</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
    <span class="p">]</span></div>



<div class="viewcode-block" id="write_proteins">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.write_proteins">[docs]</a>
<span class="k">def</span> <span class="nf">write_proteins</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">record_id</span><span class="p">,</span> <span class="n">dst_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write protein translations to a file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genes : list</span>
<span class="sd">        A list of gene objects.</span>
<span class="sd">    record_id : str</span>
<span class="sd">        The ID of the sequence record.</span>
<span class="sd">    dst_path : str</span>
<span class="sd">        The destination file path.</span>
<span class="sd">    overwrite : bool</span>
<span class="sd">        Whether to overwrite the file.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool</span>
<span class="sd">        Updated overwrite flag.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">dst_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span> <span class="k">if</span> <span class="n">overwrite</span> <span class="k">else</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">dst</span><span class="p">:</span>
        <span class="n">genes</span><span class="o">.</span><span class="n">write_translations</span><span class="p">(</span>
            <span class="n">dst</span><span class="p">,</span>
            <span class="n">sequence_id</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">record_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">width</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span>
            <span class="n">translation_table</span><span class="o">=</span><span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">translation_table</span><span class="p">,</span>
            <span class="n">include_stop</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="kc">False</span>  <span class="c1"># Update overwrite flag</span></div>



<div class="viewcode-block" id="write_nucleotides">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.write_nucleotides">[docs]</a>
<span class="k">def</span> <span class="nf">write_nucleotides</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">output_handle</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write nucleotide sequences to a file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    sequence : Bio.SeqRecord.SeqRecord</span>
<span class="sd">        The sequence record to write.</span>
<span class="sd">    output_handle : str</span>
<span class="sd">        The output file path.</span>
<span class="sd">    overwrite : bool</span>
<span class="sd">        Whether to overwrite the file.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool</span>
<span class="sd">        Updated overwrite flag.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_handle</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span> <span class="k">if</span> <span class="n">overwrite</span> <span class="k">else</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">dst</span><span class="p">:</span>
        <span class="n">write</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">dst</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="kc">False</span></div>



<div class="viewcode-block" id="select_top_structure">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.select_top_structure">[docs]</a>
<span class="k">def</span> <span class="nf">select_top_structure</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Select the top structure for each query based on the bitscore.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pandas.DataFrame</span>
<span class="sd">        A DataFrame with columns &#39;query&#39; and &#39;bits&#39;.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pandas.DataFrame</span>
<span class="sd">        A DataFrame with the top structure for each query.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">highest_bits_idx</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;query&quot;</span><span class="p">)[</span><span class="s2">&quot;bits&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()</span>
    <span class="c1"># Select those rows</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">highest_bits_idx</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">result</span></div>



<div class="viewcode-block" id="predict_orfs">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.predict_orfs">[docs]</a>
<span class="k">def</span> <span class="nf">predict_orfs</span><span class="p">(</span><span class="n">orf_finder</span><span class="p">,</span> <span class="n">seq</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find genes, compute coding capacity, and determine orientation.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    orf_finder : pyrodigal_gv.ViralGeneFinder</span>
<span class="sd">        The ORF finder object.</span>
<span class="sd">    seq : Bio.Seq.Seq</span>
<span class="sd">        The sequence to analyze.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tuple</span>
<span class="sd">        A tuple containing genes, coding capacity, orientation, and the ORF finder used.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">orf_finder</span><span class="o">.</span><span class="n">find_genes</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
    <span class="n">coding_capacity</span> <span class="o">=</span> <span class="n">calculate_coding_capacity</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
    <span class="n">orientation</span> <span class="o">=</span> <span class="n">find_orientation</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genes</span><span class="p">,</span> <span class="n">coding_capacity</span><span class="p">,</span> <span class="n">orientation</span><span class="p">,</span> <span class="n">orf_finder</span></div>



<div class="viewcode-block" id="get_lineage">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.get_lineage">[docs]</a>
<span class="k">def</span> <span class="nf">get_lineage</span><span class="p">(</span><span class="n">record_id</span><span class="p">,</span> <span class="n">taxonomy_data</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve the lineage of a given record from the taxonomy table.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    record_id : str</span>
<span class="sd">        The ID of the sequence record.</span>
<span class="sd">    taxonomy_data : pandas.DataFrame</span>
<span class="sd">        The taxonomy data table.</span>
<span class="sd">    taxdb : taxopy.TaxDb</span>
<span class="sd">        The taxonomy database.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list</span>
<span class="sd">        The lineage of the record.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">record_taxonomy</span> <span class="o">=</span> <span class="n">taxonomy_data</span><span class="p">[</span><span class="n">taxonomy_data</span><span class="p">[</span><span class="s2">&quot;contig&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">record_id</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">record_taxonomy</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>
    <span class="c1"># taxid_dict = record_taxonomy.set_index(&quot;contig&quot;)[&quot;taxid&quot;].to_dict()</span>
    <span class="n">tax</span> <span class="o">=</span> <span class="n">record_taxonomy</span><span class="p">[</span><span class="s2">&quot;taxonomy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">()</span><span class="o">.</span><span class="n">removesuffix</span><span class="p">(</span><span class="s2">&quot; sp.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">tax</span> <span class="o">==</span> <span class="s2">&quot;unclassified viruses&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">taxid</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">taxid_from_name</span><span class="p">(</span><span class="n">tax</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">)</span>
        <span class="n">lineage</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="n">taxid</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">taxdb</span><span class="p">)</span><span class="o">.</span><span class="n">name_lineage</span>
        <span class="k">return</span> <span class="n">lineage</span>
    <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
        <span class="n">click</span><span class="o">.</span><span class="n">echo</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Warning: &#39;</span><span class="si">{</span><span class="n">tax</span><span class="si">}</span><span class="s2">&#39; is not part of the official ICTV taxonomy. Its lineage can not be looked up and therefore the nucleotide reorientation could not be performed for </span><span class="si">{</span><span class="n">record_taxonomy</span><span class="p">[</span><span class="s1">&#39;contig&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">[]</span></div>



<span class="c1"># Functions to generate and save NCBI feature tables</span>
<div class="viewcode-block" id="save_ncbi_feature_tables">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.save_ncbi_feature_tables">[docs]</a>
<span class="k">def</span> <span class="nf">save_ncbi_feature_tables</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">single_file</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate and save NCBI feature tables for sequences in a DataFrame.</span>

<span class="sd">    This function creates a single feature table file by default, but can</span>
<span class="sd">    also save separate files for each unique sequence ID when specified.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        DataFrame containing sequence data with columns</span>
<span class="sd">        [&#39;seqid&#39;, &#39;accession&#39;, &#39;start&#39;, &#39;end&#39;, &#39;strand&#39;, &#39;type&#39;,</span>
<span class="sd">        &#39;Protein names&#39;, &#39;source&#39;, &#39;start_codon&#39;, &#39;partial_begin&#39;, &#39;partial_end&#39;].</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Directory path to save the feature tables. Defaults to &quot;.&quot;.</span>
<span class="sd">    single_file : bool, optional</span>
<span class="sd">        If True, saves all features to one file; otherwise, saves separate files.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">single_file</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;featuretable.tbl&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">seqid</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;seqid&quot;</span><span class="p">):</span>
                <span class="n">accession</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;seqid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;Feature </span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">write_feature_entries</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">group</span><span class="p">)</span>
        <span class="n">click</span><span class="o">.</span><span class="n">echo</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;feature_tables&quot;</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">seqid</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;seqid&quot;</span><span class="p">):</span>
            <span class="n">accession</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;seqid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;feature_tables&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">.tbl&quot;</span><span class="p">)</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
                <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;Feature </span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">write_feature_entries</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">group</span><span class="p">)</span>
            <span class="n">click</span><span class="o">.</span><span class="n">echo</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="write_feature_entries">
<a class="viewcode-back" href="../../source/suvtk.html#suvtk.features.write_feature_entries">[docs]</a>
<span class="k">def</span> <span class="nf">write_feature_entries</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">group</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function to write feature entries to a file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file : file-like object</span>
<span class="sd">        The file to write to.</span>
<span class="sd">    group : pandas.DataFrame</span>
<span class="sd">        The group of feature entries to write.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;partial_end&quot;</span><span class="p">]:</span>
            <span class="n">end</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&lt;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;partial_begin&quot;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span>
            <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>

        <span class="n">protein</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Protein names&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notna</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Protein names&quot;</span><span class="p">])</span>
            <span class="k">else</span> <span class="s2">&quot;hypothetical protein&quot;</span>
        <span class="p">)</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="s2">product</span><span class="se">\t</span><span class="si">{</span><span class="n">protein</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="s2">inference</span><span class="se">\t</span><span class="s2">ab initio prediction:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;start_codon&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;ATG&quot;</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="s2">note</span><span class="se">\t</span><span class="s2">Alternative start codon: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start_codon&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">protein</span> <span class="o">!=</span> <span class="s2">&quot;hypothetical protein&quot;</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="s2">inference</span><span class="se">\t</span><span class="s2">alignment:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;aligner&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;aligner_version&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">:UniProtKB:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Uniref_entry&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,BFVD:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;model&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span></div>



<span class="nd">@click</span><span class="o">.</span><span class="n">command</span><span class="p">(</span><span class="n">short_help</span><span class="o">=</span><span class="s2">&quot;Create feature tables for sequences.&quot;</span><span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;-i&quot;</span><span class="p">,</span>
    <span class="s2">&quot;--input&quot;</span><span class="p">,</span>
    <span class="s2">&quot;fasta_file&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">exists</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Input fasta file&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;-o&quot;</span><span class="p">,</span>
    <span class="s2">&quot;--output&quot;</span><span class="p">,</span>
    <span class="s2">&quot;output_path&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">exists</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output directory&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;-d&quot;</span><span class="p">,</span>
    <span class="s2">&quot;--database&quot;</span><span class="p">,</span>
    <span class="s2">&quot;database&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">exists</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to the suvtk database folder.&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;-g&quot;</span><span class="p">,</span>
    <span class="s2">&quot;--translation-table&quot;</span><span class="p">,</span>
    <span class="s2">&quot;transl_table&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="n">callback</span><span class="o">=</span><span class="n">validate_translation_table</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Translation table to use. Only genetic codes from https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi are allowed (1-6, 9-16, 21-31).&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;--coding-complete&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Do not predict incomplete genes (no stop codon) and only keep genomes that are &#39;coding complete&#39; (&gt;50</span><span class="si">% c</span><span class="s2">oding capacity). [This can not be turned off for now]&quot;</span><span class="p">,</span>  <span class="c1"># TODO: check on pyrodigal to implement fixed start codon</span>
<span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;--taxonomy&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="n">click</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">exists</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxonomy file to adjust sequence orientation (ssRNA- sequences will get 3&#39; -&gt; 5&#39; orientation, all others 5&#39; -&gt; 3&#39;).&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;--separate-files&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">is_flag</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Save feature tables into separate files&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nd">@click</span><span class="o">.</span><span class="n">option</span><span class="p">(</span>
    <span class="s2">&quot;-t&quot;</span><span class="p">,</span>
    <span class="s2">&quot;--threads&quot;</span><span class="p">,</span>
    <span class="s2">&quot;threads&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="n">utils</span><span class="o">.</span><span class="n">get_available_cpus</span><span class="p">(),</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of threads to use&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="k">def</span> <span class="nf">features</span><span class="p">(</span>
    <span class="n">fasta_file</span><span class="p">,</span>
    <span class="n">output_path</span><span class="p">,</span>
    <span class="n">database</span><span class="p">,</span>
    <span class="n">transl_table</span><span class="p">,</span>
    <span class="n">coding_complete</span><span class="p">,</span>
    <span class="n">taxonomy</span><span class="p">,</span>
    <span class="n">separate_files</span><span class="p">,</span>
    <span class="n">threads</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create feature tables for sequences from an input fasta file.</span>

<span class="sd">    This command processes the input sequences to predict open reading frames (ORFs),</span>
<span class="sd">    aligns the predicted protein sequences against a specified database with proteins and their function, and generates</span>
<span class="sd">    feature tables for submission to GenBank.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output_path</span><span class="p">):</span>
        <span class="n">click</span><span class="o">.</span><span class="n">echo</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Warning: Output directory &#39;</span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&#39; already exists and will be overwritten.&quot;</span>
        <span class="p">)</span>

    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>

    <span class="c1"># Train ORF finder</span>
    <span class="n">orf_finder</span> <span class="o">=</span> <span class="n">pyrodigal_gv</span><span class="o">.</span><span class="n">ViralGeneFinder</span><span class="p">()</span>
    <span class="n">training_info</span> <span class="o">=</span> <span class="n">orf_finder</span><span class="o">.</span><span class="n">train</span><span class="p">(</span>
        <span class="o">*</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">records</span><span class="p">),</span> <span class="n">translation_table</span><span class="o">=</span><span class="n">transl_table</span>
    <span class="p">)</span>

    <span class="c1"># Initialize ORF finders</span>
    <span class="n">orf_finder1</span> <span class="o">=</span> <span class="n">pyrodigal_gv</span><span class="o">.</span><span class="n">ViralGeneFinder</span><span class="p">(</span>
        <span class="n">meta</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">viral_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">closed</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">training_info</span><span class="o">=</span><span class="n">training_info</span>
    <span class="p">)</span>

    <span class="c1"># Not possible for now because pyrodigal does not support &#39;closed=[True, False]&#39; yet</span>
    <span class="c1"># orf_finder2 = pyrodigal_gv.ViralGeneFinder(</span>
    <span class="c1">#    meta=False, viral_only=True, closed=[True, False], training_info=training_info</span>
    <span class="c1"># )</span>

    <span class="c1"># Load taxonomy database</span>
    <span class="k">if</span> <span class="n">taxonomy</span><span class="p">:</span>
        <span class="n">taxdb</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">TaxDb</span><span class="p">(</span>
            <span class="n">nodes_dmp</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="s2">&quot;nodes.dmp&quot;</span><span class="p">),</span>
            <span class="n">names_dmp</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="s2">&quot;names.dmp&quot;</span><span class="p">),</span>
        <span class="p">)</span>  <span class="c1"># TODO: Set better database path?</span>
        <span class="n">taxonomy_data</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">safe_read_csv</span><span class="p">(</span><span class="n">taxonomy</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Define output paths</span>
    <span class="n">prot_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;proteins.faa&quot;</span><span class="p">)</span>
    <span class="n">nucl_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;reoriented_nucleotide_sequences.fna&quot;</span><span class="p">)</span>

    <span class="n">results</span><span class="p">,</span> <span class="n">no_orf_pred</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
    <span class="n">overwrite</span><span class="p">,</span> <span class="n">overwrite_n</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span>

    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
        <span class="n">lineage</span> <span class="o">=</span> <span class="n">get_lineage</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">taxonomy_data</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">)</span> <span class="k">if</span> <span class="n">taxonomy</span> <span class="k">else</span> <span class="p">[]</span>

        <span class="c1"># Predict ORFs using orf_finder1 first</span>
        <span class="n">genes</span><span class="p">,</span> <span class="n">coding_capacity</span><span class="p">,</span> <span class="n">orientation</span><span class="p">,</span> <span class="n">chosen_orf_finder</span> <span class="o">=</span> <span class="n">predict_orfs</span><span class="p">(</span>
            <span class="n">orf_finder1</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span>
        <span class="p">)</span>

        <span class="c1"># Commented out because it is not possible to use orf_finder2 yet</span>
        <span class="c1"># If coding capacity is too low, use orf_finder2 instead</span>
        <span class="c1"># if coding_capacity &lt; 0.5 and not coding_complete:</span>
        <span class="c1">#    # click.echo(f&quot;Repredicting ORFs for {record.id} due to low coding capacity.&quot;)</span>
        <span class="c1">#    genes, coding_capacity, orientation, chosen_orf_finder = predict_orfs(</span>
        <span class="c1">#        orf_finder2, record.seq</span>
        <span class="c1">#    )</span>

        <span class="k">if</span> <span class="n">coding_capacity</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="p">:</span>
            <span class="c1"># Adjust orientation based on lineage</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">orientation</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="s2">&quot;Negarnaviricota&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">lineage</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
                <span class="n">orientation</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="s2">&quot;Negarnaviricota&quot;</span> <span class="ow">in</span> <span class="n">lineage</span>
            <span class="p">):</span>
                <span class="n">record</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
                <span class="n">genes</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">predict_orfs</span><span class="p">(</span>
                    <span class="n">chosen_orf_finder</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span>
                <span class="p">)</span>  <span class="c1"># Use the last used ORF finder</span>

            <span class="n">results</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">extract_gene_results</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)))</span>
            <span class="n">overwrite</span> <span class="o">=</span> <span class="n">write_proteins</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">prot_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">)</span>
            <span class="n">overwrite_n</span> <span class="o">=</span> <span class="n">write_nucleotides</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">nucl_path</span><span class="p">,</span> <span class="n">overwrite_n</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">no_orf_pred</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="c1"># click.echo(</span>
            <span class="c1">#    f&quot;No ORF predictions with start site and &gt;50% coding capacity for {record.id}.&quot;</span>
            <span class="c1"># )</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;no_ORF_prediction.txt&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">no_orf_pred</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">line</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Create DataFrame from results</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;seqid&quot;</span><span class="p">,</span>
        <span class="s2">&quot;seq_length&quot;</span><span class="p">,</span>
        <span class="s2">&quot;orf&quot;</span><span class="p">,</span>
        <span class="s2">&quot;start&quot;</span><span class="p">,</span>
        <span class="s2">&quot;end&quot;</span><span class="p">,</span>
        <span class="s2">&quot;strand&quot;</span><span class="p">,</span>
        <span class="s2">&quot;start_codon&quot;</span><span class="p">,</span>
        <span class="s2">&quot;partial_begin&quot;</span><span class="p">,</span>
        <span class="s2">&quot;partial_end&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>

    <span class="n">feat_pred</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;pyrodigal-gv&quot;</span>
    <span class="n">feat_pred_version</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pyrodigal_gv</span><span class="o">.</span><span class="n">__version__</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;seqid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;seqid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;CDS&quot;</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat_pred</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">feat_pred_version</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="c1"># df[&quot;source&quot;] = f&quot;pyrodigal-gv&quot;</span>
    <span class="c1"># df[&quot;annotation_source&quot;]=f&quot;BFVD (https://doi.org/10.1093/nar/gkae1119)&quot;</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;annotation_source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;UniProtKB&quot;</span>

    <span class="c1"># Cmd = &quot;diamond blastp &quot;</span>
    <span class="c1"># Cmd += f&quot;--db {database}/foldseek_db/bfvd.dmnd &quot;</span>
    <span class="c1"># Cmd += f&quot;--query {output_path}/proteins.faa &quot;</span>
    <span class="c1"># Cmd += f&quot;--out {output_path}/alignment.m8 &quot;</span>
    <span class="c1"># Cmd += &quot;--threads {threads} &quot;</span>
    <span class="c1"># Cmd += &quot;--sensitive &quot;</span>
    <span class="c1"># Cmd += &quot;--index-chunks 1 &quot;</span>
    <span class="c1"># Cmd += &quot;--block-size 8 &quot;</span>
    <span class="c1"># Cmd += &quot;--unal 1 &quot;</span>
    <span class="c1"># Cmd += &quot;--tmpdir /dev/shm &quot;</span>
    <span class="c1"># Cmd += &quot;--outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore&quot;</span>
    <span class="c1"># utils.Exec(Cmd)</span>
    <span class="c1">#</span>
    <span class="c1"># aligner = &quot;Diamond&quot;</span>
    <span class="c1"># aligner_version = utils.Exec(&quot;diamond version&quot;, capture=True)</span>
    <span class="c1"># aligner_version = aligner_version.strip().split()[2]</span>

    <span class="n">m8_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;alignment.m8&quot;</span><span class="p">)</span>

    <span class="n">Cmd</span> <span class="o">=</span> <span class="s2">&quot;mmseqs easy-search &quot;</span>
    <span class="n">Cmd</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prot_path</span><span class="si">}</span><span class="s2"> &quot;</span>  <span class="c1"># input</span>
    <span class="n">Cmd</span> <span class="o">+=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="s2">&quot;bfvd&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span>  <span class="c1"># database</span>
    <span class="n">Cmd</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">m8_path</span><span class="si">}</span><span class="s2"> &quot;</span>  <span class="c1"># output</span>
    <span class="n">Cmd</span> <span class="o">+=</span> <span class="s2">&quot;tmp &quot;</span>  <span class="c1"># temp directory</span>
    <span class="n">Cmd</span> <span class="o">+=</span> <span class="s2">&quot;-s 7.5 &quot;</span>
    <span class="n">Cmd</span> <span class="o">+=</span> <span class="s2">&quot;--format-mode 0 &quot;</span>
    <span class="n">Cmd</span> <span class="o">+=</span> <span class="s2">&quot;--format-output query,target,pident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits &quot;</span>
    <span class="n">Cmd</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;--threads </span><span class="si">{</span><span class="n">threads</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">Exec</span><span class="p">(</span><span class="n">Cmd</span><span class="p">)</span>

    <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="s2">&quot;tmp&quot;</span><span class="p">)</span>

    <span class="n">aligner</span> <span class="o">=</span> <span class="s2">&quot;MMseqs2&quot;</span>
    <span class="n">aligner_version</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">Exec</span><span class="p">(</span><span class="s2">&quot;mmseqs version&quot;</span><span class="p">,</span> <span class="n">capture</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="n">m8</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">safe_read_csv</span><span class="p">(</span>
        <span class="n">m8_path</span><span class="p">,</span>
        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">m8</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="mi">0</span><span class="p">:</span> <span class="s2">&quot;query&quot;</span><span class="p">,</span>
            <span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;target&quot;</span><span class="p">,</span>
            <span class="mi">2</span><span class="p">:</span> <span class="s2">&quot;pident&quot;</span><span class="p">,</span>
            <span class="mi">3</span><span class="p">:</span> <span class="s2">&quot;len&quot;</span><span class="p">,</span>
            <span class="mi">4</span><span class="p">:</span> <span class="s2">&quot;mismatch&quot;</span><span class="p">,</span>
            <span class="mi">5</span><span class="p">:</span> <span class="s2">&quot;gapopen&quot;</span><span class="p">,</span>
            <span class="mi">6</span><span class="p">:</span> <span class="s2">&quot;qstart&quot;</span><span class="p">,</span>
            <span class="mi">7</span><span class="p">:</span> <span class="s2">&quot;qend&quot;</span><span class="p">,</span>
            <span class="mi">8</span><span class="p">:</span> <span class="s2">&quot;tstart&quot;</span><span class="p">,</span>
            <span class="mi">9</span><span class="p">:</span> <span class="s2">&quot;tend&quot;</span><span class="p">,</span>
            <span class="mi">10</span><span class="p">:</span> <span class="s2">&quot;evalue&quot;</span><span class="p">,</span>
            <span class="mi">11</span><span class="p">:</span> <span class="s2">&quot;bits&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">m8</span> <span class="o">=</span> <span class="n">m8</span><span class="p">[</span><span class="n">m8</span><span class="p">[</span><span class="s2">&quot;evalue&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">1e-3</span><span class="p">]</span>

    <span class="n">m8</span><span class="p">[</span><span class="s2">&quot;aligner&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aligner</span>
    <span class="n">m8</span><span class="p">[</span><span class="s2">&quot;aligner_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aligner_version</span>

    <span class="n">m8_top</span> <span class="o">=</span> <span class="n">select_top_structure</span><span class="p">(</span><span class="n">m8</span><span class="p">)</span>
    <span class="n">names_df</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">safe_read_csv</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="s2">&quot;bfvd_uniprot_names.tsv&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span>
    <span class="p">)</span>  <span class="c1"># TODO find better solution?</span>

    <span class="c1"># remove all trailing strings within brackets from protein names</span>
    <span class="n">names_df</span><span class="p">[</span><span class="s2">&quot;Protein names&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">names_df</span><span class="p">[</span><span class="s2">&quot;Protein names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;[\(\[].*?[\)\]]$&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>

    <span class="n">meta_df</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">safe_read_csv</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="s2">&quot;bfvd_metadata.tsv&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">)</span>  <span class="c1"># TODO find better solution?</span>
    <span class="n">meta_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="mi">0</span><span class="p">:</span> <span class="s2">&quot;Uniref_entry&quot;</span><span class="p">,</span>
            <span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;model&quot;</span><span class="p">,</span>
            <span class="mi">2</span><span class="p">:</span> <span class="s2">&quot;length&quot;</span><span class="p">,</span>
            <span class="mi">3</span><span class="p">:</span> <span class="s2">&quot;avg_pLDDT&quot;</span><span class="p">,</span>
            <span class="mi">4</span><span class="p">:</span> <span class="s2">&quot;pTM&quot;</span><span class="p">,</span>
            <span class="mi">5</span><span class="p">:</span> <span class="s2">&quot;splitted&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">meta_df</span><span class="p">[</span><span class="s2">&quot;model&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta_df</span><span class="p">[</span><span class="s2">&quot;model&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="n">merged_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">meta_df</span><span class="p">,</span> <span class="n">names_df</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;Uniref_entry&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;From&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span>
    <span class="p">)</span>

    <span class="n">prot_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">m8_top</span><span class="p">,</span> <span class="n">merged_df</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;target&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;model&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span>
    <span class="p">)</span>

    <span class="c1"># prot_df[&quot;Protein names&quot;]</span>

    <span class="n">prot_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;tophit_info.tsv&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">diamond</span> <span class="o">=</span> <span class="n">prot_df</span>
    <span class="n">final_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">diamond</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;orf&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;query&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>

    <span class="n">single_file</span> <span class="o">=</span> <span class="kc">False</span> <span class="k">if</span> <span class="n">separate_files</span> <span class="k">else</span> <span class="kc">True</span>
    <span class="c1"># Call the function to save feature tables</span>
    <span class="n">save_ncbi_feature_tables</span><span class="p">(</span>
        <span class="n">final_df</span><span class="p">,</span> <span class="n">output_dir</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">single_file</span><span class="o">=</span><span class="n">single_file</span>
    <span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;miuvig_features.tsv&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MIUVIG_parameter</span><span class="se">\t</span><span class="s2">value</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;feat_pred</span><span class="se">\t</span><span class="si">{</span><span class="n">feat_pred</span><span class="si">}</span><span class="s2">;</span><span class="si">{</span><span class="n">feat_pred_version</span><span class="si">}</span><span class="s2">;-g </span><span class="si">{</span><span class="n">transl_table</span><span class="si">}</span><span class="s2">, default otherwise</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;ref_db</span><span class="se">\t</span><span class="s2">BFVD;2023_02;https://bfvd.steineggerlab.workers.dev</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>  <span class="c1"># TODO: read DB version from version.txt or something</span>
        <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;sim_search_meth</span><span class="se">\t</span><span class="si">{</span><span class="n">aligner</span><span class="si">}</span><span class="s2">;</span><span class="si">{</span><span class="n">aligner_version</span><span class="si">}</span><span class="s2">;-s 7.5, default otherwise</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">features</span><span class="p">()</span>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2025, Lander De Coninck
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/LanderDC/suvtk/" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../_static/documentation_options.js?v=1b390748"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/furo.js?v=5fa4622c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    </body>
</html>